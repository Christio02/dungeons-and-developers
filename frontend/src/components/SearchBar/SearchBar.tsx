import { useEffect, useRef, useState } from 'react';
import { CiSearch } from 'react-icons/ci';

interface SearchBarProps {
  searchTerm: string;
  handleSearchChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  suggestions: string[];
  onSuggestionClick: (suggestion: string) => void;
  placeholder?: string;
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
}

const SearchBar = ({
  searchTerm,
  handleSearchChange,
  suggestions,
  onSuggestionClick,
  placeholder = 'Search for a monster...',
  onKeyDown,
}: SearchBarProps) => {
  const [isFocused, setIsFocused] = useState(false);
  const [highlightedIndex, setHighlightedIndex] = useState(-1);
  const inputRef = useRef<HTMLInputElement>(null);

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'ArrowDown') {
      setHighlightedIndex((prevIndex) => (prevIndex < suggestions.length - 1 ? prevIndex + 1 : 0));
    } else if (e.key === 'ArrowUp') {
      setHighlightedIndex((prevIndex) => (prevIndex > 0 ? prevIndex - 1 : suggestions.length - 1));
    } else if (e.key === 'Enter') {
      if (highlightedIndex >= 0) {
        onSuggestionClick(suggestions[highlightedIndex]);
        setIsFocused(false);
        setHighlightedIndex(-1);
        inputRef.current?.blur();
      } else if (onKeyDown) {
        onKeyDown(e);
      }
    }
  };

  const handleSuggestionClick = (suggestion: string) => {
    onSuggestionClick(suggestion);
    setIsFocused(false);
    setHighlightedIndex(-1);
    inputRef.current?.blur();
  };

  useEffect(() => {
    setHighlightedIndex(-1);
  }, [suggestions]);

  return (
    <div className="relative flex flex-col items-center">
      <div className="flex justify-center items-center">
        <CiSearch size="25" className="absolute left-3 top-1/2 transform -translate-y-[60%] text-gray-400" />
        <input
          ref={inputRef}
          id="search-input"
          type="text"
          placeholder={placeholder}
          maxLength={50}
          value={searchTerm}
          onChange={handleSearchChange}
          onFocus={() => setIsFocused(true)}
          onBlur={() => setTimeout(() => setIsFocused(false), 100)}
          onKeyDown={handleKeyDown}
          className="pl-12 px-[1vw] py-[0.5vh] w-[75vw] sm:w-[60vw] md:w-[50vw] xl:w-[30vw] 2xl:w-[25vw] rounded-lg border !text-black border-gray-300 focus:outline-none focus:ring-2 focus:ring-red-500 sub-header"
        />
      </div>
      {isFocused && suggestions.length > 0 && (
        <ul className="absolute top-full mt-1 w-full bg-customGray shadow-xl shadow-black rounded z-10 focus: visible">
          {suggestions.map((suggestion, index) => (
            <li
              key={index}
              onMouseDown={() => handleSuggestionClick(suggestion)}
              onMouseEnter={() => setHighlightedIndex(index)}
              className={`px-4 py-2 cursor-pointer ${
                highlightedIndex === index ? 'bg-customRed text-white' : 'hover:bg-customRed'
              } rounded`}
            >
              {suggestion}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default SearchBar;
