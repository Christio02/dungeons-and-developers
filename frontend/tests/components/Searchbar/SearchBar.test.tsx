import { render, screen } from '@testing-library/react';
import SearchBar from '../../../src/components/SearchBar/SearchBar';
import { expect, vi } from 'vitest';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';

describe('Searchbar component', () => {
  const defaultPlaceholder = 'Search for a monster...';

  const initialSearchTerm = 'Dragon';
  const handleSearchChangeMock = vi.fn();
  const mockOnSuggestionClick = vi.fn();
  const mockSuggestions = ['giant', 'dragon'];
  it('renders correctly with default placeholder and search term', () => {
    render(
      <SearchBar
        searchTerm={initialSearchTerm}
        handleSearchChange={handleSearchChangeMock}
        onSuggestionClick={mockOnSuggestionClick}
        suggestions={mockSuggestions}
      />
    );
    const inputElement = screen.getByPlaceholderText(defaultPlaceholder) as HTMLInputElement;
    expect(inputElement).toBeTruthy();
    expect(inputElement.value).toBe(initialSearchTerm);
  });

  it('calls handleSearchChange on input change', async () => {
    const user = userEvent.setup();
    render(
      <SearchBar
        searchTerm=""
        handleSearchChange={handleSearchChangeMock}
        onSuggestionClick={mockOnSuggestionClick}
        suggestions={mockSuggestions}
      />
    );
    const inputElement = screen.getByPlaceholderText(defaultPlaceholder);
    await user.type(inputElement, 'Goblin');
    expect(handleSearchChangeMock).toHaveBeenCalledTimes(6);
  });
  it('updates input value as user types', async () => {
    const user = userEvent.setup();

    const WrapperComponent = () => {
      const [searchTerm, setSearchTerm] = useState('');
      const handleSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setSearchTerm(e.target.value);
      };
      return (
        <SearchBar
          searchTerm={searchTerm}
          handleSearchChange={handleSearchChange}
          onSuggestionClick={mockOnSuggestionClick}
          suggestions={mockSuggestions}
        />
      );
    };
    render(<WrapperComponent />);
    const inputElement = screen.getByPlaceholderText(defaultPlaceholder) as HTMLInputElement;

    await user.type(inputElement, 'Orc');
    expect(inputElement.value).toBe('Orc');
  });
  it('renders the search icon', () => {
    const { container } = render(
      <SearchBar
        searchTerm={initialSearchTerm}
        handleSearchChange={handleSearchChangeMock}
        onSuggestionClick={mockOnSuggestionClick}
        suggestions={mockSuggestions}
      />
    );
    const svgElement = container.querySelector('svg');
    expect(svgElement).toBeTruthy();
  });
  it('matches snapshot', () => {
    const { container } = render(
      <SearchBar
        searchTerm={initialSearchTerm}
        handleSearchChange={handleSearchChangeMock}
        onSuggestionClick={mockOnSuggestionClick}
        suggestions={mockSuggestions}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
