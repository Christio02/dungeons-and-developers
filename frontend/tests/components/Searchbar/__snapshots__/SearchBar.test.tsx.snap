// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Searchbar component > matches snapshot 1`] = `
<div>
  <div
    class="relative flex flex-col items-center"
  >
    <div
      class="flex justify-center items-center"
    >
      <svg
        class="absolute left-3 top-1/2 transform -translate-y-[60%] text-gray-400"
        fill="currentColor"
        height="25"
        id="search"
        stroke="currentColor"
        stroke-width="0"
        version="1.1"
        viewBox="0 0 24 24"
        width="25"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <g>
          <path
            d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
          />
        </g>
      </svg>
      <input
        class="pl-12 px-[1vw] py-[0.5vh] w-[75vw] sm:w-[60vw] md:w-[50vw] xl:w-[30vw] 2xl:w-[25vw] rounded-lg border !text-black border-gray-300 focus:outline-none focus:ring-2 focus:ring-red-500 sub-header"
        maxlength="50"
        placeholder="Search for a monster..."
        type="text"
        value="Dragon"
      />
    </div>
  </div>
</div>
`;
